relay-ourlogs/src/heroku_to_sentry.rs

```diff
@@ -0,0 +1,556 @@
+
+/// Header names for Heroku Logplex integration.
+///
+/// These are the keys used to store Heroku-specific metadata on envelope items.
+/// The values come from HTTP headers sent by Logplex in HTTPS drain requests.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum HerokuHeader {
+    /// The unique request identifier.
+    ///
+    /// Logplex sets this to a unique ID for each request. If a request is retried
+    /// (e.g., due to non-2xx response or network failure), this ID enables detection
+    /// of duplicate requests.
+    FrameId,
+    /// The drain token.
+    ///
+    /// This token identifies the specific log drain sending the request.
+    /// It can be used to associate logs with their source drain configuration.
+    DrainToken,
+    /// The Logplex user agent.
+    ///
+    /// Describes the version of Logplex (e.g., "Logplex/v72").
+    /// This changes with Logplex releases.
+    UserAgent,
+}
+
+impl HerokuHeader {
+    /// Returns the HTTP header name sent by Logplex.
+    ///
+    /// These are the original header names from the Logplex HTTPS drain request.
+    pub fn http_header_name(&self) -> &'static str {
+        match self {
+            Self::FrameId => "Logplex-Frame-Id",
+            Self::DrainToken => "Logplex-Drain-Token",
+            Self::UserAgent => "User-Agent",
+        }
+    }
+
+    /// Returns the string key used to store this header on envelope items.
+    pub fn as_str(&self) -> &'static str {
+        match self {
+            Self::FrameId => "heroku-frame-id",
+            Self::DrainToken => "heroku-drain-token",
+            Self::UserAgent => "heroku-user-agent",
+        }
+    }
+}
+
+/// Parsed message body from a Logplex syslog message.
+///
+/// This struct stores the raw message body along with any logfmt key-value pairs
+/// that were parsed from it, using Sentry's message parameterization conventions.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct LogplexMsgBody<'a> {
+    /// The raw message body.
+    pub msg: &'a str,
+    /// The parameterized template string (if logfmt pairs were found).
+    /// Example: `"source={source} sample#db_size={sample#db_size}"`
+    pub template: Option<String>,
+    /// The parsed logfmt parameters as (key, value) pairs.
+    /// Keys are the original logfmt keys, values are the parsed values.
+    pub parameters: Vec<(&'a str, &'a str)>,
+}
+
+impl<'a> ParseMsg<'a> for LogplexMsgBody<'a> {
+    fn parse(msg: &'a str, _: &Originator) -> Result<(&'a str, Self), rsyslog::Error<'a>> {
+        let mut template_parts = Vec::new();
+        let mut parameters = Vec::new();
+
+        for pair in msg.split_whitespace() {
+            if let Some((key, value)) = pair.split_once('=')
+                && !key.is_empty()
+            {
+                template_parts.push(format!("{key}={{{key}}}"));
+                parameters.push((key, value));
+            }
+        }
+
+        Ok((
+            "",
```

Comment:
Is this correct? The `rsyslog` crate seems to handle the remainders differently, e.g. in `LineRaw`.